Fix incorrect `timestamp_max` max value.

batiati opened the pull request:
Since the most significant bit of the timestamp is the tombstone flag, we need to fix the `timestamp_max` as a `maxInt(u63)`.

batiati committed:
Fix incorrect `timestamp_max` max value.

batiati committed:
Validate timestamp as u63

batiati committed:
Prefer using the constant decl

sentientwaffle commented on docs/reference/account-filter.md:
```diff
@@ -27,7 +27,7 @@ Optional; set to zero to disable the lower-bound filter.
 
 - Type is 64-bit unsigned integer (8 bytes)
-- Must not be `2^64 - 1`
+- Must not overflow a 64-bit signed integer (`2^63 - 1`).
```
I don't think we should mention signed ints here... `-1` doesn't overflow a 64-bit signed int, but isn't a valid value here. Specifying a maximum value is simpler.

```suggestion
- Must be less than `2^63`.
```

(Likewise for `create_transfers.md`.)

batiati committed:
Fix assertion

Ported `self.prefetch_timestamp != TimestampRange.timestamp_max` to `self.prefetch_timestamp < TimestampRange.timestamp_max`, since it cannot be greater than timestamp_max` either. But just realized that it had no reason to use `!=` instead of `<=`.

sentientwaffle commented on src/lsm/timestamp_range.zig:
```diff
@@ -2,7 +2,7 @@ const std = @import("std");
 pub const TimestampRange = struct {
     pub const timestamp_min = 1;
-    pub const timestamp_max = std.math.maxInt(u64) - 1;
+    pub const timestamp_max = std.math.maxInt(u63); // The last bit is the tombstone flag.
```
Maybe add a note that the maximum is inclusive.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -1840,9 +1842,9 @@ pub fn StateMachineType(
-                @as(u64, t.timeout) * std.time.ns_per_s,
+                u63,
+                @intCast(t.timestamp),
+                @as(u63, t.timeout) * std.time.ns_per_s,
```
Couldn't this multiplication overflow too? (Before the sum.)

sentientwaffle replied:
Ahh right, `timeout` is only a `u32`!

batiati committed:
Review

batiati committed:
Enforce timestamp range for groove objects

batiati commented:
Sorry @sentientwaffle I added a couple more asserts after your approval!

batiati committed:
Timestamp cannot be zero anymore.
